src/evolutions/src/main/scala/play/api/db/slick/evolutions/internal/DBApiAdapter.scala

```diff
@@ -54,38 +51,71 @@ private[evolutions] object DBApiAdapter {
 
     lazy val url: String = withConnection(_.getMetaData.getURL())
 
-    def getConnection(): Connection = dbConfig.db.source.createConnection()
+    override def getConnection(): Connection = getConnection(autocommit = true)
 
-    def getConnection(autocommit: Boolean): Connection = {
-      val conn = getConnection()
-      conn.setAutoCommit(autocommit)
-      conn
+    override def getConnection(autocommit: Boolean): Connection = {
+      val connection = dbConfig.db.source.createConnection()
+      try {
+        connection.setAutoCommit(autocommit)
+      } catch {
+        case e: Throwable =>
+          // setAutoCommit can fail, so we need to close the connection
+          // which usually means we are returning it back to the pool and
+          // avoiding a leak.
+          connection.close()
+          throw e
+      }
+      connection
     }
 
     def withConnection[A](block: Connection => A): A = {
-      val conn = getConnection()
-      try block(conn)
-      finally {
-        try conn.close() catch { case _: SQLException => }
-      }
+      withConnection(autocommit = true)(block)
     }
 
-    def withConnection[A](autocommit: Boolean)(block: Connection => A): A = withConnection { conn =>
-      conn.setAutoCommit(autocommit)
-      block(conn)
+    def withConnection[A](autocommit: Boolean)(block: Connection => A): A = {
```

Comment:
Make this method the default base that other overloads of `withConnection` will use. It is responsible to execute a block and release the connection.